third_party/move/move-compiler-v2/src/ast_simplifier.rs

```diff
@@ -0,0 +1,876 @@
+        if let Some(exp) = exp {
+            match exp.as_ref() {
+                ExpData::Value(_, val) => Some(SimpleValue::Value(val.clone())),
+                ExpData::LocalVar(_id, sym) => self.values.get(sym).cloned(),
+                ExpData::Temporary(id, idx) => {
+                    if let Some(sym) = self.cached_params.get(*idx).map(|p| &p.0) {
+                        self.values.get(sym).cloned()
+                    } else {
+                        panic!(
+                            "Invalid index {} for Temporary at node {}",
+                            *idx,
+                            id.as_usize()
+                        )
+                    }
+                },
+                _ => None,
+            }
+        } else {
+            None
+        }
+    }
+
+    // Expand a `Value::Tuple` value expression to a call to `Tuple`
+    fn expand_tuple(&mut self, exp: Exp) -> Exp {
+        if let ExpData::Value(id, Value::Tuple(x)) = exp.as_ref() {
+            if x.is_empty() {
+                ExpData::Call(*id, Operation::Tuple, Vec::new()).into_exp()
+            } else {
+                let loc = self.env.get_node_loc(*id);
+                self.env.diag(
+                    Severity::Bug,
+                    &loc,
+                    &format!(
+                        "Illegal use of non-empty Tuple constant of length {}",
+                        x.len()
+                    ),
+                );
+                exp
+            }
+        } else if let ExpData::Value(id, Value::Vector(x)) = exp.as_ref() {
+            if x.is_empty() {
+                ExpData::Call(*id, Operation::Vector, Vec::new()).into_exp()
+            } else {
+                let loc = self.env.get_node_loc(*id);
+                self.env.diag(
+                    Severity::Bug,
+                    &loc,
+                    &format!(
+                        "Illegal use of non-empty Vector constant of length {}",
+                        x.len()
+                    ),
+                );
+                exp
+            }
+        } else {
+            exp
+        }
+    }
+}
+
+impl<'env> ExpRewriterFunctions for SimplifierRewriter<'env> {
+    fn rewrite_enter_scope<'a>(
+        &mut self,
+        id: NodeId,
+        vars: impl Iterator<Item = &'a (NodeId, Symbol)>,
+    ) {
+        self.visiting_binding.enter_scope();
+        self.values.enter_scope();
+        self.remapped_symbol.enter_scope();
+        for (_, sym) in vars {
+            self.visiting_binding.insert(*sym, id);
+            let sym_is_unsafe = self.unsafe_variables.contains(&(*sym, Some(id)));
+            let mapped_sym = {
+                let new_sym = if self.remapped_symbol.contains_key(sym) {
+                    // There is already a var named `sym` in the enclosing scope.
+                    // Rename this one.
+                    self.env.symbol_pool().make_unique(*sym)
```

Comment:
This is really not good because of error messages (which includes BTW the Move Prover), is there a way around this?